image-loader/app/model/ImageUpload.scala

```diff
@@ -28,57 +29,92 @@ case object ImageUpload {
+      val sourceDimensionsFuture = FileMetadataReader.dimensions(uploadedFile, uploadRequest.mimeType)
+
+      val thumbFuture = for {
+        fileMetadata <- fileMetadataFuture
+        colourModel <- colourModelFuture
+        iccColourSpace = FileMetadataHelper.normalisedIccColourSpace(fileMetadata)
+        thumb <- ImageOperations.createThumbnail(uploadedFile, thumbWidth, thumbQuality, tempDir, iccColourSpace, colourModel)
+      } yield thumb
+
+      val pngOption = if (uploadIsPng24) {
+        val optimisedPng = tempDir.getAbsolutePath() + "/optimisedpng" + ".png"
+        Seq("pngquant", "--quality", "1-85", uploadedFile.getAbsolutePath(), "--output", optimisedPng).!
+
+        Some(new File(optimisedPng))
+      } else
+        None
+
+      bracket(thumbFuture)(_.delete) { thumb =>
+        // Run the operations in parallel
+        val thumbStoreFuture = storeThumbnail(uploadRequest, thumb)
+
+        val pngStoreFuture = if (uploadIsPng24)
+            Some(storeOptimisedPng(uploadRequest, pngOption.get)).map(result => result.map(Option(_)))
+              .getOrElse(Future.successful(None))
+            else
+              Future(None)
+
+        val thumbDimensionsFuture = FileMetadataReader.dimensions(thumb, Some("image/jpeg"))
+
+        val pngDimensionsFuture = if (uploadIsPng24)
+            FileMetadataReader.dimensions(pngOption.get, Some("image/png"))
+          else
+            Future(None)
+
+        for {
+          s3Source <- sourceStoreFuture
+          s3Thumb <- thumbStoreFuture
+          s3PngOption <- pngStoreFuture
+          sourceDimensions <- sourceDimensionsFuture
+          thumbDimensions <- thumbDimensionsFuture
+          pngDimensions <- pngDimensionsFuture
+          fileMetadata <- fileMetadataFuture
+          colourModel <- colourModelFuture
+          fullFileMetadata = fileMetadata.copy(colourModel = colourModel)
+
+          metadata = ImageMetadataConverter.fromFileMetadata(fullFileMetadata)
+          cleanMetadata = metadataCleaners.clean(metadata)
+
+          sourceAsset = Asset.fromS3Object(s3Source, sourceDimensions)
+          thumbAsset = Asset.fromS3Object(s3Thumb, thumbDimensions)
+          pngAsset = if (uploadIsPng24)
+            Some(Asset.fromS3Object(s3PngOption.get, pngDimensions))
+          else
+            None
+
+          baseImage = createImage(uploadRequest, sourceAsset, thumbAsset, pngAsset, fullFileMetadata, cleanMetadata)
+          processedImage = SupplierProcessors.process(baseImage)
+
+          // FIXME: dirty hack to sync the originalUsageRights and originalMetadata as well
+          finalImage = processedImage.copy(
+            originalMetadata = processedImage.metadata,
+            originalUsageRights = processedImage.usageRights
+          )
+        }
+          yield {
+            ImageUpload(uploadRequest, finalImage)
+          }
       }
     }
-  }
+  }.flatMap(f => f)
```

Comment:
This seems shady, I'm surprised you need it, maybe it's because the "yield" in a "for" comprehension calls "map" whereas the arrows call "flatMap", so perhaps adding another flattening before the yield... not sure though, happy to sit with you and try it out.